consensus/src/consensus_observer/observer.rs

```diff
@@ -166,28 +168,51 @@ impl ConsensusObserver {
         debug!(LogSchema::new(LogEntry::ConsensusObserver)
             .message("Checking consensus observer progress!"));
 
-        // Get the peer ID of the currently active subscription
+        // Get the peer ID of the currently active subscription (if any)
         let active_subscription_peer = self
             .active_observer_subscription
             .as_ref()
             .map(|subscription| subscription.get_peer_network_id());
 
         // If we have an active subscription, verify that the subscription
-        // is still healthy. If not, the subscription will be terminated.
-        if active_subscription_peer.is_some() {
-            self.check_active_subscription().await;
+        // is still healthy. If not, the subscription should be terminated.
+        if let Some(active_subscription_peer) = active_subscription_peer {
+            if let Err(error) = self.check_active_subscription().await {
+                // Log the subscription termination
+                warn!(
+                    LogSchema::new(LogEntry::ConsensusObserver).message(&format!(
+                        "Terminating subscription to peer: {:?}! Error: {:?}",
+                        active_subscription_peer, error
+                    ))
+                );
+
+                // Unsubscribe from the peer
+                self.unsubscribe_from_peer(active_subscription_peer);
+
+                // Update the subscription termination metrics
+                self.update_subscription_termination_metrics(active_subscription_peer, error);
+            }
         }
 
-        // If we no longer have an active subscription, select a new peer to subscribe to
+        // If we don't have a subscription, we should select a new peer to
+        // subscribe to. If we had a previous subscription, it should be
+        // excluded from the selection process.
         if self.active_observer_subscription.is_none() {
+            // Create a new observer subscription
             self.create_new_observer_subscription(active_subscription_peer)
```

Comment:
this is a blocking rpc call, maybe we should have some metric for the latency

Reply:
Yeah, we already track RPC latencies (see: https://github.com/aptos-labs/aptos-core/blob/main/consensus/src/consensus_observer/network_client.rs#L212). I can also add a metric for this function call (if we want), as it may make several blocking RPCs in a row until it can find a peer 😄